sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -556,6 +556,109 @@ Parameters:
 
 |===
 
+==== Dynamic Command Groups
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+class dynamic_command_group {
+public:
+  dynamic_command_group(
+      command_graph<graph_state::modifiable> graph, const std::vector<std::function<void(handler &)>>& cgfList);
+
+  size_t get_active_cgf();
+  void set_active_cgf(size_t cgfIndex);
+};
+----
+
+Dynamic command-groups can be added as nodes to a graph. They provide a mechanism that
+allows updating the command-group function of a node after the graph is finalized.
+There is always one command-group function in the dynamic command-group that is set
+as active. When a dynamic command-group node is executed, the kernel of the active
+command-group function will be run and all the other command-group functions in
+`cgfList` will be ignored.
+
+See <<executable-graph-update, Executable Graph Update>> for more information
+about updating command-groups.
+
+===== Limitations
+
+Dynamic command-groups can only be used to update kernels. Trying to update a command group
+function that contains other operations will result in an error.
+
+All the command-group functions in a dynamic command-group must have identical dependencies.
+It is not allowed for a dynamic command-group to have command-group functions that would
+result in a change to the graph topology when set to active. In practice, this means that
+any calls to `handler.depends_on()` must be identical for all the command-group functions
+in a dynamic command-group.
+
+Table {counter: tableNumber}. Member functions of the `dynamic_command_group` class.
+[cols="2a,a"]
+|===
+|Member Function|Description
+
+|
+[source,c++]
+----
+dynamic_command_group(
+command_graph<graph_state::modifiable> graph,
+const std::vector<std::function<void(handler &)>>& cgfList);
+----
+
+|Constructs a dynamic command-group object that can be added as a node to a `command_graph``.
+
+Parameters:
+
+* `graph` - Graph to be associated with this `dynamic_command_group`.
+* `cgfList` - The list of command-group functions that can be activated for this dynamic command-group.
+              The command-group function at index 0 will be active by default.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the graph wasn't created with
+  the `property::graph::assume_buffer_outlives_graph` property and the `dynamic_command_group`
+  is created with command-group functions that use buffers. See the
+  <<assume-buffer-outlives-graph-property, Assume-Buffer-Outlives-Graph>>
+  property for more information.
+
+* Throws with error code `invalid` if the `dynamic_command_group` is created with
+  command-group functions that are not kernel executions.
```

Comment:
I guess if it really is too complicated, I can move this error to graph.add()

But on the other hand this also feels like the right place to do this kind of error checking. Otherwise we will end up with the invalid dynamic_command_group object.

The handler doesn't seem to be doing anything special before handler::finalize() is called. So  could we potentially initialize it here and call finalize() in the graph?

It would be a bit of an issue if the same dynamic command-group is used in 2 different nodes though. Some logic to create a new handler might be needed in that case.

Reply:
The `dynamic_command_group` constructor has a parameter for the modifiable graph, so we could at least construct a handler with `handler(std::shared_ptr<ext::oneapi::experimental::detail::graph_impl> Graph);` and  then do `CGF(Handler); Handler.getType()` If we never used the handler after that and it was destroyed i'm not sure if that would be a problem.

Regardless, we probably need to actually try implement this to work things out. So I'd suggest specifying the ideal behavior for now (which is an error here), and then refining it once we have tried to implement it.

Reply:
> Regardless, we probably need to actually try implement this to work things out. So I'd suggest specifying the ideal behavior for now (which is an error here), and then refining it once we have tried to implement it.

Makes sense, I'm fine with that.